libvmaf/src/feature/funque_vif.c

```diff
@@ -0,0 +1,241 @@
+    y_pad = (funque_dtype*)malloc(sizeof(funque_dtype) * (width + (2 * y_reflect)) * (height + (2 * y_reflect)));
+
+    reflect_pad(x, width, height, x_reflect, x_pad);
+    reflect_pad(y, width, height, y_reflect, y_pad);
+    size_t r_width = width + (2 * x_reflect);
+    size_t r_height = height + (2 * x_reflect);
+
+    double* int_1_x, * int_1_y, * int_2_x, * int_2_y, * int_xy;
+    double* var_x, * var_y, * cov_xy;
+
+    int_1_x = (double*)calloc((r_width + 1) * (r_height + 1), sizeof(double));
+    int_1_y = (double*)calloc((r_width + 1) * (r_height + 1), sizeof(double));
+    int_2_x = (double*)calloc((r_width + 1) * (r_height + 1), sizeof(double));
+    int_2_y = (double*)calloc((r_width + 1) * (r_height + 1), sizeof(double));
+    int_xy = (double*)calloc((r_width + 1) * (r_height + 1), sizeof(double));
+
+    integral_image(x_pad, r_width, r_height, int_1_x);
+    integral_image(y_pad, r_width, r_height, int_1_y);
+    integral_image_2(x_pad, x_pad, r_width, r_height, int_2_x);
+    integral_image_2(y_pad, y_pad, r_width, r_height, int_2_y);
+    integral_image_2(x_pad, y_pad, r_width, r_height, int_xy);
+
+    var_x = (double*)malloc(sizeof(double) * (r_width + 1 - kw) * (r_height + 1 - kh));
+    var_y = (double*)malloc(sizeof(double) * (r_width + 1 - kw) * (r_height + 1 - kh));
+    cov_xy = (double*)malloc(sizeof(double) * (r_width + 1 - kw) * (r_height + 1 - kh));
+
+    compute_metrics(int_1_x, int_1_y, int_2_x, int_2_y, int_xy, r_width + 1, r_height + 1, kh, kw, k_norm, var_x, var_y, cov_xy);
+
+    size_t s_width = (r_width + 1) - kw;
+    size_t s_height = (r_height + 1) - kh;
+
+    double* g = (double*)malloc(sizeof(double) * s_width * s_height);
+    double* sv_sq = (double*)malloc(sizeof(double) * s_width * s_height);
+    double exp = (double)1e-10;
+    int index;
+
+    *score = (double)0;
+    *score_num = (double)0;
+    *score_den = (double)0;
+
+    for (unsigned int i = 0; i < s_height; i++)
+    {
+        for (unsigned int j = 0; j < s_width; j++)
+        {
+            index = i * s_width + j;
+
+            g[index] = cov_xy[index] / (var_x[index] + exp);
+            sv_sq[index] = var_y[index] - g[index] * cov_xy[index];
+
+            if (var_x[index] < exp)
+            {
+                g[index] = (double)0;
+                sv_sq[index] = var_y[index];
+                var_x[index] = (double)0;
+            }
+
+            if (var_y[index] < exp)
+            {
+                g[index] = (double)0;
+                sv_sq[index] = (double)0;
+            }
+
+            if (g[index] < 0)
+            {
+                sv_sq[index] = var_x[index];
+                g[index] = (double)0;
+            }
+
+            if (sv_sq[index] < exp)
+                sv_sq[index] = exp;
+
+            *score_num += (log((double)1 + g[index] * g[index] * var_x[index] / (sv_sq[index] + sigma_nsq)) + (double)1e-4);
+            *score_den += (log((double)1 + var_x[index] / sigma_nsq) + (double)1e-4);
```

Comment:
You can add a preprocessor flag to switch between the 1e-4 method (used here) and VMAF's method https://github.com/Netflix/vmaf/blob/master/libvmaf/src/feature/vif_tools.c#L369-376. A comparison between the two would be great.